In [1]:
import pandas as pd


portfolio1 = pd.read_csv('data/estaticos_portfolio1.csv',index_col=0)
portfolio2 = pd.read_csv('data/estaticos_portfolio2.csv',index_col=0)
portfolio3 = pd.read_csv('data/estaticos_portfolio2.csv',index_col=0)
#market = pd.read_csv('data/estaticos_market.csv',index_col=0)


portfolio2.head() ,portfolio3.head()

(                                                  id
 0  09e95c1a84049001d086470a8f320a19b076f955a89122...
 1  dc9d155f5bcd3172031e2ec9431786e8871ff4b0ff4728...
 2  16843c9ffb920170477118933798510d8d5f00f5c56c12...
 3  ff045934d90aab184dd15e66daf3b3c97193bda5449298...
 4  a0ce528caab2c62497a548d0ebd9e935143ef81ed7dbf0...,
                                                   id
 0  09e95c1a84049001d086470a8f320a19b076f955a89122...
 1  dc9d155f5bcd3172031e2ec9431786e8871ff4b0ff4728...
 2  16843c9ffb920170477118933798510d8d5f00f5c56c12...
 3  ff045934d90aab184dd15e66daf3b3c97193bda5449298...
 4  a0ce528caab2c62497a548d0ebd9e935143ef81ed7dbf0...)

portfolio2 and portfolio3 are simply composed of ids, so the exploratory analysis will be done on portfolio1

In [2]:
portfolio1.head()

id  fl_matriz  \
0  dabe79bec87c88ae04e869bf6bd321ee5e1893cecf6625...       True   
1  32e5f4e10932153a7ba869cb0386e7e02d49d2461046b8...      False   
2  a95d6f30bba445bd3d6b0c5b36f865b38ec01d17336090...       True   
3  6cb309685cea0b6d2988818792ec2e6fcb2bd02e0afa9e...       True   
4  f72aa7fa6787b0a5a1c88885b6120850df8ee0f71adc25...       True   

            de_natureza_juridica sg_uf natureza_juridica_macro  \
0          EMPRESARIO INDIVIDUAL    PI                  OUTROS   
1          EMPRESARIO INDIVIDUAL    MA                  OUTROS   
2          EMPRESARIO INDIVIDUAL    MA                  OUTROS   
3          EMPRESARIO INDIVIDUAL    RN                  OUTROS   
4  SOCIEDADE EMPRESARIA LIMITADA    RN  ENTIDADES EMPRESARIAIS   

              de_ramo     setor  idade_empresa_anos idade_emp_cat  fl_me  ...  \
0  COMERCIO VAREJISTA  COMERCIO            0.646575          <= 1  False  ...   
1  COMERCIO VAREJISTA  COMERCIO            6.586301        5 a 10  False  ...   
2  COMERCIO VAREJISTA  COMERCIO            8.010959        5 a 10  False  ...   
3  COMERCIO VAREJISTA  COMERCIO           20.863014          > 20  False  ...   
4  COMERCIO VAREJISTA  COMERCIO           18.726027       15 a 20  False  ...   

   media_meses_servicos  max_meses_servicos  min_meses_servicos  \
0                   NaN                 NaN                 NaN   
1                   NaN                 NaN                 NaN   
2                   NaN                 NaN                 NaN   
3                   NaN                 NaN                 NaN   
4             39.454762                91.2           19.333333   

   qt_funcionarios qt_funcionarios_12meses  qt_funcionarios_24meses  \
0              NaN                     NaN                      NaN   
1              0.0                     0.0                      2.0   
2              NaN                     NaN                      NaN   
3              NaN                     NaN                      NaN   
4             14.0                    15.0                     25.0   

   tx_crescimento_12meses  tx_crescimento_24meses tx_rotatividade qt_filiais  
0                     NaN                     NaN             NaN          0  
1                     NaN                  -100.0             0.0          4  
2                     NaN                     NaN             NaN          0  
3                     NaN                     NaN             NaN          0  
4               -6.666667                   -44.0             0.0          3  

[5 rows x 181 columns]

In [3]:
#number of rows and columns respectively
portfolio1.shape

(555, 181)

In [4]:
portfolio1.describe()

idade_empresa_anos  vl_total_tancagem  vl_total_veiculos_antt  \
count          555.000000                0.0                     0.0   
mean            10.139247                NaN                     NaN   
std              9.641625                NaN                     NaN   
min              0.052055                NaN                     NaN   
25%              2.924658                NaN                     NaN   
50%              6.709589                NaN                     NaN   
75%             15.494521                NaN                     NaN   
max             47.517808                NaN                     NaN   

       vl_total_veiculos_leves  vl_total_veiculos_pesados    qt_art  \
count                38.000000                  38.000000  4.000000   
mean                  1.473684                   0.526316  3.250000   
std                   1.309773                   1.330248  3.201562   
min                   0.000000                   0.000000  1.000000   
25%                   1.000000                   0.000000  1.750000   
50%                   1.000000                   0.000000  2.000000   
75%                   2.000000                   0.750000  3.500000   
max                   6.000000                   7.000000  8.000000   

       vl_total_veiculos_pesados_grupo  vl_total_veiculos_leves_grupo  \
count                       555.000000                     555.000000   
mean                          0.861261                       0.542342   
std                          16.312180                       8.502044   
min                           0.000000                       0.000000   
25%                           0.000000                       0.000000   
50%                           0.000000                       0.000000   
75%                           0.000000                       0.000000   
max                         382.000000                     199.000000   

       vl_total_tancagem_grupo  vl_total_veiculos_antt_grupo  ...  \
count                      0.0                           0.0  ...   
mean                       NaN                           NaN  ...   
std                        NaN                           NaN  ...   
min                        NaN                           NaN  ...   
25%                        NaN                           NaN  ...   
50%                        NaN                           NaN  ...   
75%                        NaN                           NaN  ...   
max                        NaN                           NaN  ...   

       media_meses_servicos  max_meses_servicos  min_meses_servicos  \
count            101.000000          101.000000          101.000000   
mean              49.929006           72.989769           33.615512   
std               40.968949           61.443495           37.410508   
min                5.406667            5.800000            2.166667   
25%               22.323077           25.233333            5.966667   
50%               41.966667           54.700000           22.133333   
75%               65.916667           93.266667           48.566667   
max              228.155556          309.333333          192.633333   

       qt_funcionarios  qt_funcionarios_12meses  qt_funcionarios_24meses  \
count       131.000000               131.000000               131.000000   
mean          4.458015                 4.679389                 4.908397   
std           9.053057                 9.875585                 9.396767   
min           0.000000                 0.000000                 0.000000   
25%           1.000000                 0.000000                 0.000000   
50%           1.000000                 1.000000                 2.000000   
75%           4.000000                 4.500000                 5.000000   
max          70.000000                75.000000                74.000000   

       tx_crescimento_12meses  tx_crescimento_24meses  tx_rotatividade  \
count               97.000000          

In [5]:
#list of features
list(portfolio1.columns)

['id',
 'fl_matriz',
 'de_natureza_juridica',
 'sg_uf',
 'natureza_juridica_macro',
 'de_ramo',
 'setor',
 'idade_empresa_anos',
 'idade_emp_cat',
 'fl_me',
 'fl_sa',
 'fl_epp',
 'fl_mei',
 'fl_ltda',
 'dt_situacao',
 'fl_st_especial',
 'fl_email',
 'fl_telefone',
 'fl_rm',
 'nm_divisao',
 'nm_segmento',
 'fl_spa',
 'fl_antt',
 'fl_veiculo',
 'vl_total_tancagem',
 'vl_total_veiculos_antt',
 'vl_total_veiculos_leves',
 'vl_total_veiculos_pesados',
 'fl_optante_simples',
 'qt_art',
 'vl_total_veiculos_pesados_grupo',
 'vl_total_veiculos_leves_grupo',
 'vl_total_tancagem_grupo',
 'vl_total_veiculos_antt_grupo',
 'vl_potenc_cons_oleo_gas',
 'fl_optante_simei',
 'sg_uf_matriz',
 'de_saude_tributaria',
 'de_saude_rescencia',
 'nu_meses_rescencia',
 'de_nivel_atividade',
 'de_indicador_telefone',
 'fl_simples_irregular',
 'vl_frota',
 'empsetorcensitariofaixarendapopulacao',
 'nm_meso_regiao',
 'nm_micro_regiao',
 'fl_passivel_iss',
 'qt_socios',
 'qt_socios_pf',
 'qt_socios_pj',
 'idade_medi

In [6]:
#number of columns with more than 75% of the values missing
most_missing_cols = set(portfolio1.loc[:, portfolio1.isna().mean() > .75])
len(most_missing_cols)

127

In [7]:
#dataframe composed of the columns with less than 75% of the values missing
less75 = portfolio1.drop(columns=most_missing_cols)
less75.head()

id  fl_matriz  \
0  dabe79bec87c88ae04e869bf6bd321ee5e1893cecf6625...       True   
1  32e5f4e10932153a7ba869cb0386e7e02d49d2461046b8...      False   
2  a95d6f30bba445bd3d6b0c5b36f865b38ec01d17336090...       True   
3  6cb309685cea0b6d2988818792ec2e6fcb2bd02e0afa9e...       True   
4  f72aa7fa6787b0a5a1c88885b6120850df8ee0f71adc25...       True   

            de_natureza_juridica sg_uf natureza_juridica_macro  \
0          EMPRESARIO INDIVIDUAL    PI                  OUTROS   
1          EMPRESARIO INDIVIDUAL    MA                  OUTROS   
2          EMPRESARIO INDIVIDUAL    MA                  OUTROS   
3          EMPRESARIO INDIVIDUAL    RN                  OUTROS   
4  SOCIEDADE EMPRESARIA LIMITADA    RN  ENTIDADES EMPRESARIAIS   

              de_ramo     setor  idade_empresa_anos idade_emp_cat  fl_me  ...  \
0  COMERCIO VAREJISTA  COMERCIO            0.646575          <= 1  False  ...   
1  COMERCIO VAREJISTA  COMERCIO            6.586301        5 a 10  False  ...   
2  COMERCIO VAREJISTA  COMERCIO            8.010959        5 a 10  False  ...   
3  COMERCIO VAREJISTA  COMERCIO           20.863014          > 20  False  ...   
4  COMERCIO VAREJISTA  COMERCIO           18.726027       15 a 20  False  ...   

   qt_socios_st_regular  qt_socios_masculino  qt_socios_feminino  \
0                   1.0                  1.0                 NaN   
1                   NaN                  NaN                 NaN   
2                   1.0                  NaN                 1.0   
3                   1.0                  1.0                 NaN   
4                   1.0                  NaN                 1.0   

          de_faixa_faturamento_estimado  \
0       DE R$ 81.000,01 A R$ 360.000,00   
1       DE R$ 81.000,01 A R$ 360.000,00   
2                      ATE R$ 81.000,00   
3                      ATE R$ 81.000,00   
4  DE R$ 1.500.000,01 A R$ 4.800.000,00   

     de_faixa_faturamento_estimado_grupo  vl_faturamento_estimado_aux  \
0        DE R$ 81.000,01 A R$ 360.000,00                     210000.0   
1     DE R$ 360.000,01 A R$ 1.500.000,00                     210000.0   
2                       ATE R$ 81.000,00                      50000.0   
3                       ATE R$ 81.000,00                      50000.0   
4  DE R$ 4.800.000,01 A R$ 10.000.000,00                    3029140.8   

   vl_faturamento_estimado_grupo_aux  qt_ex_funcionarios  \
0                           210000.0                 NaN   
1                          1000915.2                 6.0   
2                            50000.0                 NaN   
3                            50000.0                 NaN   
4                          5872824.0                28.0   

  qt_funcionarios_grupo qt_filiais  
0                   NaN          0  
1                   2.0          4  
2                   NaN          0  
3                   NaN          0  
4                  28.0          3  

[5 rows x 54 columns]

In [8]:
# percentage of values still missing in the remainder of the dataframe
less75.isna().mean().sort_values(ascending=False)

qt_ex_funcionarios                       0.736937
qt_funcionarios_grupo                    0.736937
qt_socios_feminino                       0.657658
qt_socios_masculino                      0.643243
qt_socios_st_regular                     0.351351
idade_maxima_socios                      0.347748
idade_media_socios                       0.347748
idade_minima_socios                      0.347748
empsetorcensitariofaixarendapopulacao    0.270270
qt_socios_pf                             0.264865
qt_socios_pj                             0.264865
qt_socios                                0.264865
fl_optante_simples                       0.181982
fl_optante_simei                         0.181982
nm_micro_regiao                          0.115315
nm_meso_regiao                           0.115315
nu_meses_rescencia                       0.084685
de_faixa_faturamento_estimado            0.045045
vl_faturamento_estimado_aux              0.045045
vl_faturamento_estimado_grupo_aux        0.045045


In [9]:
#There are still a couple of columns with a high % of values missing, so let's restrict it to 50%
less50_cols = set(portfolio1.loc[:, portfolio1.isna().mean() > .5])
less50 = portfolio1.drop(columns=less50_cols)

less50.isna().mean().sort_values(ascending=False)

qt_socios_st_regular                     0.351351
idade_minima_socios                      0.347748
idade_maxima_socios                      0.347748
idade_media_socios                       0.347748
empsetorcensitariofaixarendapopulacao    0.270270
qt_socios_pj                             0.264865
qt_socios_pf                             0.264865
qt_socios                                0.264865
fl_optante_simei                         0.181982
fl_optante_simples                       0.181982
nm_micro_regiao                          0.115315
nm_meso_regiao                           0.115315
nu_meses_rescencia                       0.084685
de_faixa_faturamento_estimado            0.045045
vl_faturamento_estimado_grupo_aux        0.045045
de_faixa_faturamento_estimado_grupo      0.045045
vl_faturamento_estimado_aux              0.045045
de_saude_rescencia                       0.019820
de_saude_tributaria                      0.019820
de_nivel_atividade                       0.007207


In [10]:
#Now the columns with highest value missing count is at 35, let's explore it further and decide what to do with it later on

df = less50 #to simplify the exploration let's just rename our dataframe to df
df.describe()

idade_empresa_anos  vl_total_veiculos_pesados_grupo  \
count          555.000000                       555.000000   
mean            10.139247                         0.861261   
std              9.641625                        16.312180   
min              0.052055                         0.000000   
25%              2.924658                         0.000000   
50%              6.709589                         0.000000   
75%             15.494521                         0.000000   
max             47.517808                       382.000000   

       vl_total_veiculos_leves_grupo  nu_meses_rescencia  \
count                     555.000000          508.000000   
mean                        0.542342           24.771654   
std                         8.502044            9.969729   
min                         0.000000            7.000000   
25%                         0.000000           22.000000   
50%                         0.000000           23.000000   
75%                         0.000000           25.000000   
max                       199.000000           54.000000   

       empsetorcensitariofaixarendapopulacao   qt_socios  qt_socios_pf  \
count                             405.000000  408.000000    408.000000   
mean                             1130.532765    1.279412      1.272059   
std                               815.038203    0.654125      0.647811   
min                               110.300000    1.000000      1.000000   
25%                               643.070000    1.000000      1.000000   
50%                               905.940000    1.000000      1.000000   
75%                              1303.820000    1.000000      1.000000   
max                              6201.910000    9.000000      9.000000   

       qt_socios_pj  idade_media_socios  idade_maxima_socios  \
count    408.000000          362.000000           362.000000   
mean       0.007353           41.753223            42.872928   
std        0.110593           12.647311            13.486646   
min        0.000000            7.500000             9.000000   
25%        0.000000           32.000000            32.000000   
50%        0.000000           40.000000            41.000000   
75%        0.000000           50.000000            51.000000   
max        2.000000           81.000000            83.000000   

       idade_minima_socios  qt_socios_st_regular  vl_faturamento_estimado_aux  \
count           362.000000            360.000000                 5.300000e+02   
mean             40.690608              1.194444                 3.218504e+05   
std              12.682938              0.501776                 1.005342e+06   
min               6.000000              1.000000                 0.000000e+00   
25%              31.000000              1.000000                 5.000000e+04   
50%              40.000000              1.000000                 2.100000e+05   
75%              49.000000              1.000000                 2.100000e+05   
max              80.000000              6.000000                 2.019427e+07   

       vl_faturamento_estimado_grupo_aux  qt_filiais  
count                       5.300000e+02  555.000000  
mean                        6.359388e+06    1.236036  
std                         1.165307e+08   16.288026  
min                         5.000000e+04    0.000000  
25%                         5.000000e+04    0.000000  
50%                         2.100000e+05    0.000000  
75%                         2.100000e+05    0.000000  
max                         2.660064e+09  374.000000

In [11]:
#qt_socios_pj seems to have very low variability let's check it out
df.qt_socios_pj.value_counts()

0.0    406
2.0      1
1.0      1
Name: qt_socios_pj, dtype: int64

In [12]:
#according to the feature dictionary qt_socios_pf is also related to feature above so let's see the sum of unique values as well
df.qt_socios_pf.value_counts()

1.0    317
2.0     80
3.0      8
5.0      1
9.0      1
4.0      1
Name: qt_socios_pf, dtype: int64

In [13]:
#Same as above
df.qt_socios.value_counts()

1.0    315
2.0     81
3.0      9
5.0      1
9.0      1
4.0      1
Name: qt_socios, dtype: int64

These are clearly features we can transform in a single feature, since it is only stating if that partner is a registered as a company or as a person, and the number count matches the amount of partners feature (qt_socios), this means we can drop the remaing two.

In [14]:
df = df.drop(columns=['qt_socios_pf','qt_socios_pj'])

In [15]:
#checking the column types
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 555 entries, 0 to 554
Data columns (total 48 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   id                                     555 non-null    object 
 1   fl_matriz                              555 non-null    bool   
 2   de_natureza_juridica                   555 non-null    object 
 3   sg_uf                                  555 non-null    object 
 4   natureza_juridica_macro                555 non-null    object 
 5   de_ramo                                555 non-null    object 
 6   setor                                  555 non-null    object 
 7   idade_empresa_anos                     555 non-null    float64
 8   idade_emp_cat                          555 non-null    object 
 9   fl_me                                  555 non-null    bool   
 10  fl_sa                                  555 non-null    bool   
 11  fl_epp

In [16]:
#separating the numerical values from the categorical ones
num_vars = ['idade_empresa_anos','vl_total_veiculos_pesados_grupo','vl_total_veiculos_leves_grupo','nu_meses_rescencia','empsetorcensitariofaixarendapopulacao',
           'qt_socios','idade_media_socios','idade_maxima_socios','idade_minima_socios','qt_socios_st_regular',
            'vl_faturamento_estimado_aux','vl_faturamento_estimado_grupo_aux','qt_filiais']

#according to the feature dictionary some of the above features are intervals, 
#but for example idade_emp_cat is the range of idade_empresa_anos so some work is needed to simplify this

df_num_vars = df[num_vars]
df_num_vars.head()

idade_empresa_anos  vl_total_veiculos_pesados_grupo  \
0            0.646575                                0   
1            6.586301                                0   
2            8.010959                                0   
3           20.863014                                0   
4           18.726027                                0   

   vl_total_veiculos_leves_grupo  nu_meses_rescencia  \
0                              0                 NaN   
1                              0                22.0   
2                              0                24.0   
3                              0                26.0   
4                              0                23.0   

   empsetorcensitariofaixarendapopulacao  qt_socios  idade_media_socios  \
0                                 379.96        1.0                45.0   
1                                1865.61        NaN                 NaN   
2                                 798.12        1.0                54.0   
3                                 608.47        1.0                40.0   
4                                 880.71        2.0                33.0   

   idade_maxima_socios  idade_minima_socios  qt_socios_st_regular  \
0                 45.0                 45.0                   1.0   
1                  NaN                  NaN                   NaN   
2                 54.0                 54.0                   1.0   
3                 40.0                 40.0                   1.0   
4                 33.0                 33.0                   1.0   

   vl_faturamento_estimado_aux  vl_faturamento_estimado_grupo_aux  qt_filiais  
0                     210000.0                           210000.0           0  
1                     210000.0                          1000915.2           4  
2                      50000.0                            50000.0           0  
3                      50000.0                            50000.0           0  
4                    3029140.8                          5872824.0           3

In [17]:
#checking if there are repeated ids
df.id.value_counts().sort_values(ascending=False)

10778c83ac2a8035d217a7c6ea87616f31adc7b876ebb872f9be975d7e878fe1    1
216489d05aa54ecb7f426937618d0256ed7074fdbc1375138e9e9fab0ce18f66    1
4c565b049273bf6e5021624c5f14508a80e1b4caf72481d3d15e8342bfe34b1f    1
823049e3568d9d97941706ee3233d9402a41f36de5e15835394b033e8411b397    1
81c469ef072ec5687f828d3cd939bab56ecde0ba9ffacbf0cd3143c980c2f799    1
                                                                   ..
68d59e0d490f829027feca49524633f4a7a9c7fea8c551f6237119c19880a102    1
0a05ef343496299e9f8c54cd951796c3f5a13dee93aa078c6d0cb9342960480f    1
4c43f4afaf4b1d5b08f83ecbac8128b8113a9c526afe61028a8108dbaa615211    1
9714f7f05a541c0d0d14853b6468a78f3adffe2fdcc8c58046c29ebdbee300fd    1
a05a86ad99266c82bbb2a7019826ce2d8e64569cc3b4f5bf8fb4467f9bb8fe66    1
Name: id, Length: 555, dtype: int64

In [18]:
df.shape

(555, 48)

In [19]:
#taking a look if the are any rows with more than 50% of the values missing (which is 24, half of the shape above)
df.isna().sum(axis=1).sort_values(ascending=False)

235    15
469    14
15     13
232    12
253    11
       ..
171     0
170     0
169     0
168     0
554     0
Length: 555, dtype: int64

In theory if all row values are the same, then we are looking at the same id. To overule the fact that an unique id could be stored twice, let's drop the id column then check for duplicate rows.

In [20]:
df_no_ids = df.drop(columns=['id'])
df_no_ids.shape

(555, 47)

In [21]:
df_no_duplicates = df_no_ids.drop_duplicates()
df_no_duplicates.shape

(555, 47)

The number of rows stays the same so the initial hypothesys that an unique id is in fact unique is true.

These features all relate to the type of the company:


fl_me: boolean value, true if the company has the term 'ME' in the end of its legal name.

fl_sa: boolean value, true if the company has the term 'SA' in the end of its legal name.

fl_epp: boolean value, true if the company has the term 'EPP' in the end of its legal name.

fl_mei: boolean value, true if the company has the term 'MEI' in the end of its legal name.

fl_ltda: boolean value, true if the company has the term 'LTDA' in the end of its legal name.


MEI - 1 partner 1 employee

ME - micro company - 1 partner, 9 employees or 19 if industry or construction related

EPP - small size company- 1 or more + partners, 10 to 49 employees, 20 to 99 if industry or construction 

Ltda - there are no limits (partners come together)

SA - there are no limits, anonimous society (capitals come together) - usually envolve more complex structures than Ltda


Since these features describe quite well the size of the company and are already a boolean, transforming them into 0/1 pairs is all that needs to be done

In [22]:
transform_boolean = lambda col: 1 if col == True else 0 #quick function to transform True into 1 and False into 0

company_size_cols = ['fl_me','fl_sa','fl_epp','fl_mei','fl_ltda']

for col in company_size_cols:
    df[col] = df.loc[:,col].apply(transform_boolean)
    
df[company_size_cols].head()

fl_me  fl_sa  fl_epp  fl_mei  fl_ltda
0      0      0       0       1        0
1      0      0       0       0        0
2      0      0       0       1        0
3      0      0       0       0        0
4      0      0       0       0        0

Moving on to the the features that relate to the number of partners a company has as well as the average, maximum and minimum ages of those partners.

Since the age of the partners might hide some hints about the profile of the company it is interesting that we keep these features, at least for now.

As you can see below there are a couple of Nan's so let's see what the mean and median look like as well as the % of Nans.

In [23]:
df[['idade_media_socios','idade_maxima_socios','idade_minima_socios','qt_socios']]

idade_media_socios  idade_maxima_socios  idade_minima_socios  qt_socios
0                  45.0                 45.0                 45.0        1.0
1                   NaN                  NaN                  NaN        NaN
2                  54.0                 54.0                 54.0        1.0
3                  40.0                 40.0                 40.0        1.0
4                  33.0                 33.0                 33.0        2.0
..                  ...                  ...                  ...        ...
550                 NaN                  NaN                  NaN        NaN
551                40.0                 40.0                 40.0        1.0
552                 NaN                  NaN                  NaN        NaN
553                34.0                 34.0                 34.0        1.0
554                45.0                 45.0                 45.0        1.0

[555 rows x 4 columns]

In [24]:
df[['idade_media_socios','idade_maxima_socios','idade_minima_socios','qt_socios']].mean()

idade_media_socios     41.753223
idade_maxima_socios    42.872928
idade_minima_socios    40.690608
qt_socios               1.279412
dtype: float64

In [25]:
df[['idade_media_socios','idade_maxima_socios','idade_minima_socios','qt_socios']].median()

idade_media_socios     40.0
idade_maxima_socios    41.0
idade_minima_socios    40.0
qt_socios               1.0
dtype: float64

In [26]:
df[['idade_media_socios','idade_maxima_socios','idade_minima_socios','qt_socios']].isna().mean()

idade_media_socios     0.347748
idade_maxima_socios    0.347748
idade_minima_socios    0.347748
qt_socios              0.264865
dtype: float64

The median won't take into account outliers and the we get integers, which fits the pattern of the features 
so we'll fill the Nan values with it.

In [27]:
def apply_median(df,columns):
    '''takes in a list of columns from a dataframe (df) and applies the median of that column where the value is a NaN'''
    for col in columns:
        df.loc[:,col] = df.loc[:,col].fillna(df.loc[:,col].median())
    return df;

        
partner_info_cols = ['idade_media_socios','idade_maxima_socios','idade_minima_socios','qt_socios']
apply_median(df,partner_info_cols)
df[partner_info_cols].head()

idade_media_socios  idade_maxima_socios  idade_minima_socios  qt_socios
0                45.0                 45.0                 45.0        1.0
1                40.0                 41.0                 40.0        1.0
2                54.0                 54.0                 54.0        1.0
3                40.0                 40.0                 40.0        1.0
4                33.0                 33.0                 33.0        2.0

Next let's inspect the following features:


fl_matriz - it tells if the company is the headquarters or not, since the goal is to find potential customers, this doesn't relate to the problem so we can drop it 


natureza_juridica_macro - character, juridic nature of the company. Again this doen't relate to our objective, so we'll drop it too.


As for the following we'll inspect them below:
de_natureza_juridica                   
sg_uf                                   
de_ramo                                
setor                                  
idade_empresa_anos                     
idade_emp_cat  

In [28]:
df['de_natureza_juridica'].value_counts()

EMPRESARIO INDIVIDUAL                                                     432
SOCIEDADE EMPRESARIA LIMITADA                                             105
EMPRESA INDIVIDUAL DE RESPONSABILIDADE LIMITADA DE NATUREZA EMPRESARIA     15
SOCIEDADE ANONIMA FECHADA                                                   1
SOCIEDADE EM CONTA DE PARTICIPACAO                                          1
SOCIEDADE EMPRESARIA EM NOME COLETIVO                                       1
Name: de_natureza_juridica, dtype: int64

This feature is describing indirectly the size of the company however it doesn't offer a clear division like the ones we handled above, and it is the same info. Again we can drop this.

In [29]:
df['natureza_juridica_macro'].value_counts()

OUTROS                    447
ENTIDADES EMPRESARIAIS    108
Name: natureza_juridica_macro, dtype: int64

In [30]:
df['de_ramo'].value_counts()

COMERCIO VAREJISTA    496
BENS DE CONSUMO        59
Name: de_ramo, dtype: int64

In [31]:
df['setor'].value_counts()

COMERCIO     496
INDUSTRIA     52
SERVIÇO        7
Name: setor, dtype: int64

Both features relate to the sector of the company, however the second offers a more clear distinction as well as allowing better scalability. So we'll drop 'de_ramo' and transform the second one with get dummies.

In [32]:
df[['idade_empresa_anos','idade_emp_cat']]

idade_empresa_anos idade_emp_cat
0              0.646575          <= 1
1              6.586301        5 a 10
2              8.010959        5 a 10
3             20.863014          > 20
4             18.726027       15 a 20
..                  ...           ...
550           27.660274          > 20
551            1.597260         1 a 5
552           31.868493          > 20
553            0.969863          <= 1
554            3.994521         1 a 5

[555 rows x 2 columns]

Both of these features describe the age of the company.

First things first the numerical value is more descriptive since it doesn't put a 20 year old company in the same bracket as a 31, as we can see above. 

Second of all, there is something suspicious about how the numerical value of the age is being colected since it is a float with many decimal cases, maybe it is the number of days that the company it is registered, we should if possible confirmate this hypothesis.

Finally what we'll do is drop the age interval column, and round up the age to 2 decimal cases(just so that we don't categorize companies with only 5months old as the same as 1.4 years old). Another option would be ignoring the numerical values and working with labels like very new, new, recent, medium old as 0,1 values corresponding to the age intervals of the company. 


In [33]:
df = df.drop(columns=['fl_matriz','natureza_juridica_macro','de_natureza_juridica','de_ramo','idade_emp_cat'])



In [34]:
def create_dummy_df(df, cat_cols, dummy_na):
    '''
    INPUT:
    df - pandas dataframe with categorical variables you want to dummy
    cat_cols - list of strings that are associated with names of the categorical columns
    dummy_na - Bool holding whether you want to dummy NA vals of categorical columns or not
    
    OUTPUT:
    df - a new dataframe that has the following characteristics:
            1. contains all columns that were not specified as categorical
            2. removes all the original columns in cat_cols
            3. dummy columns for each of the categorical columns in cat_cols
            4. if dummy_na is True - it also contains dummy columns for the NaN values
            5. Use a prefix of the column name with an underscore (_) for separating 
    '''
    for col in  cat_cols:
        try:
            # for each cat add dummy var, drop original column
            df = pd.concat([df.drop(columns=col, axis=1), pd.get_dummies(df[col], prefix=col, prefix_sep='_', drop_first=True, dummy_na=dummy_na)], axis=1)
        except:
            continue
    return df;

#sg_uf is the state where the company if from, we'll transform into a dummy column as well.
dummy_cols = ['sg_uf','setor']
df = create_dummy_df(df,dummy_cols,False)
df.head()

id  idade_empresa_anos  \
0  dabe79bec87c88ae04e869bf6bd321ee5e1893cecf6625...            0.646575   
1  32e5f4e10932153a7ba869cb0386e7e02d49d2461046b8...            6.586301   
2  a95d6f30bba445bd3d6b0c5b36f865b38ec01d17336090...            8.010959   
3  6cb309685cea0b6d2988818792ec2e6fcb2bd02e0afa9e...           20.863014   
4  f72aa7fa6787b0a5a1c88885b6120850df8ee0f71adc25...           18.726027   

   fl_me  fl_sa  fl_epp  fl_mei  fl_ltda dt_situacao  fl_st_especial  \
0      0      0       0       1        0  2018-03-06           False   
1      0      0       0       0        0  2012-03-29           False   
2      0      0       0       1        0  2010-10-26           False   
3      0      0       0       0        0  2005-11-03           False   
4      0      0       0       0        0  2005-11-03           False   

   fl_email  ...  vl_faturamento_estimado_aux  \
0     False  ...                     210000.0   
1      True  ...                     210000.0   
2     False  ...                      50000.0   
3     False  ...                      50000.0   
4      True  ...                    3029140.8   

  vl_faturamento_estimado_grupo_aux qt_filiais sg_uf_AM  sg_uf_MA  sg_uf_PI  \
0                          210000.0          0        0         0         1   
1                         1000915.2          4        0         1         0   
2                           50000.0          0        0         1         0   
3                           50000.0          0        0         0         0   
4                         5872824.0          3        0         0         0   

   sg_uf_RN sg_uf_RO  setor_INDUSTRIA  setor_SERVIÇO  
0         0        0                0              0  
1         0        0                0              0  
2         0        0                0              0  
3         1        0                0              0  
4         1        0                0              0  

[5 rows x 48 columns]

In [35]:
df.loc[:,'idade_empresa_anos'] = df.loc[:,'idade_empresa_anos'].round(decimals=2)
df.head()

id  idade_empresa_anos  \
0  dabe79bec87c88ae04e869bf6bd321ee5e1893cecf6625...                0.65   
1  32e5f4e10932153a7ba869cb0386e7e02d49d2461046b8...                6.59   
2  a95d6f30bba445bd3d6b0c5b36f865b38ec01d17336090...                8.01   
3  6cb309685cea0b6d2988818792ec2e6fcb2bd02e0afa9e...               20.86   
4  f72aa7fa6787b0a5a1c88885b6120850df8ee0f71adc25...               18.73   

   fl_me  fl_sa  fl_epp  fl_mei  fl_ltda dt_situacao  fl_st_especial  \
0      0      0       0       1        0  2018-03-06           False   
1      0      0       0       0        0  2012-03-29           False   
2      0      0       0       1        0  2010-10-26           False   
3      0      0       0       0        0  2005-11-03           False   
4      0      0       0       0        0  2005-11-03           False   

   fl_email  ...  vl_faturamento_estimado_aux  \
0     False  ...                     210000.0   
1      True  ...                     210000.0   
2     False  ...                      50000.0   
3     False  ...                      50000.0   
4      True  ...                    3029140.8   

  vl_faturamento_estimado_grupo_aux qt_filiais sg_uf_AM  sg_uf_MA  sg_uf_PI  \
0                          210000.0          0        0         0         1   
1                         1000915.2          4        0         1         0   
2                           50000.0          0        0         1         0   
3                           50000.0          0        0         0         0   
4                         5872824.0          3        0         0         0   

   sg_uf_RN sg_uf_RO  setor_INDUSTRIA  setor_SERVIÇO  
0         0        0                0              0  
1         0        0                0              0  
2         0        0                0              0  
3         1        0                0              0  
4         1        0                0              0  

[5 rows x 48 columns]

Moving on to the next batch of features:

dt_situacao - date when the "de_situacao" was registered by the IRS. This column 'de_situacao' doesnt exist in the original dataframe as you can see below, so we'll also drop this column

fl_st_especial  - boolean value, true if dt_situacao_especial is not null. If it is not null means that some
extraordinary situation is identified by the IRS (ESPOLIO DE EMPRESARIO EMPRESA INDIVIDUAL OU
EIRELI, FALIDO, EM LIQUIDACAO, LIQUIDACAO JUDICIAL, LIQUIDACAO EXTRA JUDICIAL, REGISTRO NA
JUNTA COMERCIAL EM ANDAMENTO, EM LIQUID EXTRA JUDICIAL, RECUPERACAO JUDICIAL,
INTERVENCAO). 

This could be potentially interesting because it could tell us if the company is under some sort of financial trouble, however it's broader than that since it contains some other possibilities when the value of the feature it's true. Adding that to the fact that we have no way of separting these possibilites, it is better just to drop this feature.

Yet, on a side note if the ecomical health of the potential client is of crucial importance we might consider adding it back in the future.


fl_email  and fl_telefone indicate whether or not the company has registered emails and phones with the IRS. This only of bureaucratic importance, which means we can drop it. Arguabally it could potentially indicate how much the company is connected, but it seems farfetched at this point. If we are interested in this specific point we'll revise it in the future.


fl_rm - if the company is situated in a metropolitan area, we'll apply the transform boolean function to this as well.

nm_divisao, nm_segmento  - these relate to the activity of the company, we'll explore this below


fl_spa  - true if the company has it's own fuel station - could be a hint for companies who relate to this sector so we'll keep it and transform as above.


fl_antt - true if the company is certified by ANTT for trasnportation purposes - this means that the company somehow is related to transportation so again we'll keep it and transform.

fl_veiculo - true if the company owns at least one vehicle - could also be a hint for companies related to vehicles or even insurances, for example, so again we'll keep it and transform.


fl_optante_simples - true if the company is taxed according to SIMPLES regime - This means the company is under special taxes regime, which is somehow related to the size, since we already have features describing the size and this one in particular has NaN values we'll drop it.  

In [36]:
portfolio1['de_situacao']

KeyError: 'de_situacao'

In [37]:
df['nm_divisao'].value_counts()

COMERCIO VAREJISTA                                              496
FABRICACAO DE PRODUTOS ALIMENTICIOS                              16
FABRICACAO DE PRODUTOS DE MADEIRA                                 9
FABRICACAO DE MOVEIS                                              9
CONFECCAO DE ARTIGOS DO VESTUARIO E ACESSORIOS                    9
FABRICACAO DE PRODUTOS DIVERSOS                                   8
MANUTENCAO REPARACAO E INSTALACAO DE MAQUINAS E EQUIPAMENTOS      7
FABRICACAO DE BEBIDAS                                             1
Name: nm_divisao, dtype: int64

In [38]:
df['nm_segmento'].value_counts()

COMERCIO; REPARACAO DE VEICULOS AUTOMOTORES E MOTOCICLETAS    496
INDUSTRIAS DE TRANSFORMACAO                                    59
Name: nm_segmento, dtype: int64

The information on these two features overlaps, since it is better described in the 'nm_divisao' we'll keep that one and transform it into dummies.

In [39]:
df = df.drop(columns=['dt_situacao','fl_st_especial','fl_email','fl_telefone','nm_segmento','fl_optante_simples'])
             
df.head()

id  idade_empresa_anos  \
0  dabe79bec87c88ae04e869bf6bd321ee5e1893cecf6625...                0.65   
1  32e5f4e10932153a7ba869cb0386e7e02d49d2461046b8...                6.59   
2  a95d6f30bba445bd3d6b0c5b36f865b38ec01d17336090...                8.01   
3  6cb309685cea0b6d2988818792ec2e6fcb2bd02e0afa9e...               20.86   
4  f72aa7fa6787b0a5a1c88885b6120850df8ee0f71adc25...               18.73   

   fl_me  fl_sa  fl_epp  fl_mei  fl_ltda fl_rm          nm_divisao  fl_spa  \
0      0      0       0       1        0   NAO  COMERCIO VAREJISTA   False   
1      0      0       0       0        0   NAO  COMERCIO VAREJISTA   False   
2      0      0       0       1        0   SIM  COMERCIO VAREJISTA   False   
3      0      0       0       0        0   NAO  COMERCIO VAREJISTA   False   
4      0      0       0       0        0   SIM  COMERCIO VAREJISTA   False   

   ...  vl_faturamento_estimado_aux  vl_faturamento_estimado_grupo_aux  \
0  ...                     210000.0                           210000.0   
1  ...                     210000.0                          1000915.2   
2  ...                      50000.0                            50000.0   
3  ...                      50000.0                            50000.0   
4  ...                    3029140.8                          5872824.0   

   qt_filiais  sg_uf_AM sg_uf_MA sg_uf_PI sg_uf_RN sg_uf_RO  setor_INDUSTRIA  \
0           0         0        0        1        0        0                0   
1           4         0        1        0        0        0                0   
2           0         0        1        0        0        0                0   
3           0         0        0        0        1        0                0   
4           3         0        0        0        1        0                0   

  setor_SERVIÇO  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 42 columns]

In [40]:
cols_transform = ['fl_rm','fl_spa','fl_antt','fl_veiculo']

for col in cols_transform:
    df[col] = df.loc[:,col].apply(transform_boolean)
    
df[cols_transform].head()

fl_rm  fl_spa  fl_antt  fl_veiculo
0      0       0        0           0
1      0       0        0           0
2      0       0        0           0
3      0       0        0           0
4      0       0        0           0

In [41]:
dummy_cols2 = ['nm_divisao']
df = create_dummy_df(df,dummy_cols2,False)
df.head()

id  idade_empresa_anos  \
0  dabe79bec87c88ae04e869bf6bd321ee5e1893cecf6625...                0.65   
1  32e5f4e10932153a7ba869cb0386e7e02d49d2461046b8...                6.59   
2  a95d6f30bba445bd3d6b0c5b36f865b38ec01d17336090...                8.01   
3  6cb309685cea0b6d2988818792ec2e6fcb2bd02e0afa9e...               20.86   
4  f72aa7fa6787b0a5a1c88885b6120850df8ee0f71adc25...               18.73   

   fl_me  fl_sa  fl_epp  fl_mei  fl_ltda  fl_rm  fl_spa  fl_antt  ...  \
0      0      0       0       1        0      0       0        0  ...   
1      0      0       0       0        0      0       0        0  ...   
2      0      0       0       1        0      0       0        0  ...   
3      0      0       0       0        0      0       0        0  ...   
4      0      0       0       0        0      0       0        0  ...   

   sg_uf_RO  setor_INDUSTRIA  setor_SERVIÇO  \
0         0                0              0   
1         0                0              0   
2         0                0              0   
3         0                0              0   
4         0                0              0   

  nm_divisao_CONFECCAO DE ARTIGOS DO VESTUARIO E ACESSORIOS  \
0                                                  0          
1                                                  0          
2                                                  0          
3                                                  0          
4                                                  0          

  nm_divisao_FABRICACAO DE BEBIDAS nm_divisao_FABRICACAO DE MOVEIS  \
0                                0                               0   
1                                0                               0   
2                                0                               0   
3                                0                               0   
4                                0                               0   

  nm_divisao_FABRICACAO DE PRODUTOS ALIMENTICIOS  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   nm_divisao_FABRICACAO DE PRODUTOS DE MADEIRA  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   

  nm_divisao_FABRICACAO DE PRODUTOS DIVERSOS  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   nm_divisao_MANUTENCAO REPARACAO E INSTALACAO DE MAQUINAS E EQUIPAMENTOS  
0                                                  0                        
1                                                  0                        
2                                                  0                        
3                                                  0                        
4                                                  0                        

[5 rows x 48 columns]

vl_total_veiculos_pesados_grupo, vl_total_veiculos_leves_grupo - see below

fl_optante_simei - true if the company is taxed as an MEI - Individual Micro Entrepreneur - we already have this info the type of company, where there are no NaNs, so we can drop this column.

sg_uf_matriz - this feature tells us in which state the headquarters of the company is located, as we already have the location of the company and for the same reasons stated above we'll drop it as well.

de_saude_tributaria                    544 non-null    object 

de_saude_rescencia                     544 non-null    object 

nu_meses_rescencia                     508 non-null    float64(unfinished)

These 3 features give us information about the tax health of the company, the first is the categorical indicator, the 2nd is a categorical indicator of how long the first is delayed and the third is a numerical indicator of that delay. All of them have Nan values. 

For the indicator itself we can apply the function we have above for dummy columns for categorical data however with the dummy_na as True, having a feature that representing that we have no info about this indicator for that particular company.

As for the delay we have more categorical info than numerical, so we can transform it into a set of 0,1 features per year 0-1, +1, +2, etc where we have the numerical data for it, complement that with categorical that and finally set the remaining values to 0 where we have no data.

de_nivel_atividade                     551 non-null    object 

This feature is a categorical indicator about the probability of the company being active, since we have some NaN values, we'll apply the dummy col funtion with the NaN option as true as well.

fl_simples_irregular - true if the company is taxed under the SIMPLES regime and has a
impeditive CNAE (economic activities that are not allowed to be taxed according to the simples regime)
revenue above the limit of the regime.

This could be an indicator of growth for the company, or even willingness to abide by regulations, since there are no nan values we'll apply the transform bolean function to this feature.


empsetorcensitariofaixarendapopulacao  405 non-null    float64(unfinished)

This feature is the numerical value of the average salaray of the population. Since we have NaN values it makes sense to categorize these into intervals and have an extra column for NaN values.

nm_meso_regiao                         491 non-null    object 
nm_micro_regiao                        491 non-null    object 

The a meso_regiao is composed by a group of a micro_regiao, unless the goal of a company is to target locally there isn't much point in adding sparse columns to our data. So we'll drop the second and apply the dummy column function with the NA as true for this feature as well.

In [42]:
df['vl_total_veiculos_pesados_grupo'].value_counts()

0      535
1        9
8        2
7        2
4        2
2        2
382      1
40       1
5        1
Name: vl_total_veiculos_pesados_grupo, dtype: int64

In [43]:
df['vl_total_veiculos_leves_grupo'].value_counts()

0      513
1       23
3        7
2        5
6        2
199      1
17       1
10       1
5        1
4        1
Name: vl_total_veiculos_leves_grupo, dtype: int64

When we look at the dictionary there is a column 'vl_total_veiculos_leves' which is the total number of light weight vehicles of the company, which would be fitting to the values we are seeing above. 

While the 'vl_total_veiculos_leves_grupo' is the economic value of the light weight vehicles so we would be expecting higher values in the thousands interval. Is this represented as 1k, 10k? 

We have nothing to confirm this hyptothesis. Also there is an outlier which we can see below that it is the same id. The same happens for 'veiculos_pesados' which relates to the heavy weight vehicles.

Since theses features have no null values we could eliminate the outlier and normalize the data so it becomes representative.

However instead of making assumptions about the data what we can infer from here is if the company has light weight vehicles or heavy weight vehicles or both. So we'll transform these into 0,1 columns where 1 will be where there is a vehicle of that type.

In [44]:
df[df['vl_total_veiculos_leves_grupo']==199]

id  idade_empresa_anos  \
167  32a26c8394fa0c08967b5e1c2696cb4fd7452b0762ed94...               12.34   

     fl_me  fl_sa  fl_epp  fl_mei  fl_ltda  fl_rm  fl_spa  fl_antt  ...  \
167      0      0       0       0        0      0       0        0  ...   

     sg_uf_RO  setor_INDUSTRIA  setor_SERVIÇO  \
167         0                0              0   

    nm_divisao_CONFECCAO DE ARTIGOS DO VESTUARIO E ACESSORIOS  \
167                                                  0          

    nm_divisao_FABRICACAO DE BEBIDAS nm_divisao_FABRICACAO DE MOVEIS  \
167                                0                               0   

    nm_divisao_FABRICACAO DE PRODUTOS ALIMENTICIOS  \
167                                              0   

     nm_divisao_FABRICACAO DE PRODUTOS DE MADEIRA  \
167                                             0   

    nm_divisao_FABRICACAO DE PRODUTOS DIVERSOS  \
167                                          0   

     nm_divisao_MANUTENCAO REPARACAO E INSTALACAO DE MAQUINAS E EQUIPAMENTOS  
167                                                  0                        

[1 rows x 48 columns]

In [45]:
df[df['vl_total_veiculos_pesados_grupo']==382]

id  idade_empresa_anos  \
167  32a26c8394fa0c08967b5e1c2696cb4fd7452b0762ed94...               12.34   

     fl_me  fl_sa  fl_epp  fl_mei  fl_ltda  fl_rm  fl_spa  fl_antt  ...  \
167      0      0       0       0        0      0       0        0  ...   

     sg_uf_RO  setor_INDUSTRIA  setor_SERVIÇO  \
167         0                0              0   

    nm_divisao_CONFECCAO DE ARTIGOS DO VESTUARIO E ACESSORIOS  \
167                                                  0          

    nm_divisao_FABRICACAO DE BEBIDAS nm_divisao_FABRICACAO DE MOVEIS  \
167                                0                               0   

    nm_divisao_FABRICACAO DE PRODUTOS ALIMENTICIOS  \
167                                              0   

     nm_divisao_FABRICACAO DE PRODUTOS DE MADEIRA  \
167                                             0   

    nm_divisao_FABRICACAO DE PRODUTOS DIVERSOS  \
167                                          0   

     nm_divisao_MANUTENCAO REPARACAO E INSTALACAO DE MAQUINAS E EQUIPAMENTOS  
167                                                  0                        

[1 rows x 48 columns]

In [46]:
transform_vehicles = lambda col: 0 if col == 0 else 1 #quick function to transform having vehicles into 1 and not having vehicles into 0

vehicle_cols = ['vl_total_veiculos_leves_grupo','vl_total_veiculos_pesados_grupo']

for col in vehicle_cols:
    df[col] = df.loc[:,col].apply(transform_vehicles)
    
df[vehicle_cols].head()

vl_total_veiculos_leves_grupo  vl_total_veiculos_pesados_grupo
0                              0                                0
1                              0                                0
2                              0                                0
3                              0                                0
4                              0                                0

In [47]:
df['vl_total_veiculos_leves_grupo'].value_counts()

0    513
1     42
Name: vl_total_veiculos_leves_grupo, dtype: int64

In [48]:
df['vl_total_veiculos_pesados_grupo'].value_counts()

0    535
1     20
Name: vl_total_veiculos_pesados_grupo, dtype: int64

In [49]:
df['de_saude_tributaria'].value_counts()

VERDE       167
AZUL        128
AMARELO     113
CINZA        86
LARANJA      47
VERMELHO      3
Name: de_saude_tributaria, dtype: int64

In [50]:
df['de_saude_rescencia'].value_counts()

ACIMA DE 1 ANO    453
ATE 1 ANO          55
SEM INFORMACAO     36
Name: de_saude_rescencia, dtype: int64

In [51]:
df['nu_meses_rescencia'].value_counts()

23.0    124
22.0    107
24.0     51
48.0     35
25.0     35
26.0     28
21.0     22
27.0     16
9.0      13
10.0     13
50.0     11
8.0      11
7.0      10
47.0      6
11.0      5
19.0      5
49.0      4
28.0      3
54.0      3
12.0      3
52.0      1
40.0      1
20.0      1
Name: nu_meses_rescencia, dtype: int64

In [52]:
df['de_nivel_atividade'].value_counts() 

MEDIA          280
ALTA           153
BAIXA          116
MUITO BAIXA      2
Name: de_nivel_atividade, dtype: int64

In [53]:
df['empsetorcensitariofaixarendapopulacao']

0       379.96
1      1865.61
2       798.12
3       608.47
4       880.71
        ...   
550    1585.53
551     552.94
552        NaN
553     606.29
554    6201.91
Name: empsetorcensitariofaixarendapopulacao, Length: 555, dtype: float64

In [54]:
df['nm_meso_regiao'].value_counts()

CENTRO AMAZONENSE         78
LESTE POTIGUAR            77
NORTE MARANHENSE          65
CENTRO NORTE PIAUIENSE    55
OESTE MARANHENSE          36
LESTE MARANHENSE          25
CENTRO MARANHENSE         21
VALE DO ACRE              20
OESTE POTIGUAR            18
SUDESTE PIAUIENSE         17
NORTE PIAUIENSE           16
SUDOESTE PIAUIENSE        15
CENTRAL POTIGUAR          12
SUL MARANHENSE            10
AGRESTE POTIGUAR           9
SUL AMAZONENSE             8
SUDOESTE AMAZONENSE        6
NORTE AMAZONENSE           2
VALE DO JURUA              1
Name: nm_meso_regiao, dtype: int64

In [55]:
df['nm_micro_regiao'].value_counts()

MANAUS                            65
NATAL                             58
AGLOMERACAO URBANA DE SAO LUIS    48
TERESINA                          41
IMPERATRIZ                        20
                                  ..
SERRA DE SAO MIGUEL                1
TEFE                               1
ALTO PARNAIBA PIAUIENSE            1
CHAPADAS DAS MANGABEIRAS           1
TARAUACA                           1
Name: nm_micro_regiao, Length: 66, dtype: int64

In [56]:
dummy_col_NA_True = ['de_saude_tributaria','de_nivel_atividade','nm_meso_regiao']
df = create_dummy_df(df,dummy_col_NA_True,True)

df.head()

id  idade_empresa_anos  \
0  dabe79bec87c88ae04e869bf6bd321ee5e1893cecf6625...                0.65   
1  32e5f4e10932153a7ba869cb0386e7e02d49d2461046b8...                6.59   
2  a95d6f30bba445bd3d6b0c5b36f865b38ec01d17336090...                8.01   
3  6cb309685cea0b6d2988818792ec2e6fcb2bd02e0afa9e...               20.86   
4  f72aa7fa6787b0a5a1c88885b6120850df8ee0f71adc25...               18.73   

   fl_me  fl_sa  fl_epp  fl_mei  fl_ltda  fl_rm  fl_spa  fl_antt  ...  \
0      0      0       0       1        0      0       0        0  ...   
1      0      0       0       0        0      0       0        0  ...   
2      0      0       0       1        0      0       0        0  ...   
3      0      0       0       0        0      0       0        0  ...   
4      0      0       0       0        0      0       0        0  ...   

   nm_meso_regiao_OESTE MARANHENSE  nm_meso_regiao_OESTE POTIGUAR  \
0                                0                              0   
1                                0                              0   
2                                0                              0   
3                                0                              1   
4                                0                              0   

   nm_meso_regiao_SUDESTE PIAUIENSE nm_meso_regiao_SUDOESTE AMAZONENSE  \
0                                 0                                  0   
1                                 0                                  0   
2                                 0                                  0   
3                                 0                                  0   
4                                 0                                  0   

  nm_meso_regiao_SUDOESTE PIAUIENSE nm_meso_regiao_SUL AMAZONENSE  \
0                                 1                             0   
1                                 0                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   

   nm_meso_regiao_SUL MARANHENSE  nm_meso_regiao_VALE DO ACRE  \
0                              0                            0   
1                              0                            0   
2                              0                            0   
3                              0                            0   
4                              0                            0   

   nm_meso_regiao_VALE DO JURUA nm_meso_regiao_nan  
0                             0                  0  
1                             0                  0  
2                             0                  0  
3                             0                  0  
4                             0                  0  

[5 rows x 74 columns]

In [57]:
transform_boolean = lambda col: 1 if col == True else 0 #quick function to transform True into 1 and False into 0

cols = ['fl_simples_irregular']

for col in cols:
    df[col] = df.loc[:,col].apply(transform_boolean)
    
df[cols].head()

fl_simples_irregular
0                     0
1                     0
2                     0
3                     0
4                     0

fl_passivel_iss                        555 non-null    bool   
qt_socios_st_regular                   360 non-null    float64
de_faixa_faturamento_estimado          530 non-null    object 
de_faixa_faturamento_estimado_grupo    530 non-null    object 
vl_faturamento_estimado_aux            530 non-null    float64
vl_faturamento_estimado_grupo_aux      530 non-null    float64
qt_filiais                             555 non-null    int64